[SofaSimulationCore] More generic DefaultVisualManagerLoop

damienmarchal opened the pull request:
Currently:
  - there must be a single VisualLoop in the scene
  - this single one must be at the root of the scene. 
  - there is no way to specify or change what is the rendered node
   
This PR allows to:
 - have multiple (possibly different) visual loop in the scene (by updating simulation::draw()). 
 - let user select which node to render (by using a SingleLink instead of a c++ attribute to store the node).
 
 This is useful as it allows to increase the separation of the different "concerns" of an object and it make explicit in the GUI the link between a node and the rendered. 

Old scenes are still possible:
 ```xml
    <DefaultVisualManagerLoop"/>
    <Node name="modelling">
         <Node name="Tripod">
         </Node>
    </Node>    
 ```

 But now we can also do:
 ```xml
    <DefaultVisualManagerLoop targetNode="@/modelling/Tripod"/>

    <Node name="modelling">
         <Node name="Tripod">
         </Node>
         <Node name="Nothing-to-Display">
         </Node>
    </Node>    
 ```
 Or even
 ```xml
    <Node name="modelling">
         <Node name="Tripod">
         </Node>
    </Node>
    
    <Node name="Simulation">
    </Node>

    <Node name="Rendering">
          <DefaultVisualManagerLoop targetNode="@/modelling/Tripod"/>
    </Node>    
 ```

Visual example:
![screen](https://user-images.githubusercontent.com/12644882/146540665-19f952c3-4a9f-4add-af88-47ca9fe2c96b.png)


______________________________________________________

By submitting this pull request, I acknowledge that  
**I have read, understand, and agree [SOFA Developer Certificate of Origin (DCO)](https://github.com/sofa-framework/sofa/blob/master/CONTRIBUTING.md#sofa-developer-certificate-of-origin-dco)**.
______________________________________________________

**Reviewers will merge this pull-request only if**  
- it builds with SUCCESS for all platforms on the CI.
- it does not generate new warnings.
- it does not generate new unit test failures.
- it does not generate new scene test failures.
- it does not break API compatibility.
- it is more than 1 week old (or has fast-merge label).

damienmarchal committed:
[SofaKernel] A simulation can now handle multiple VisualLoop.

damienmarchal committed:
[SofaKernel] Use SingleLink to store the rendered target

This allows the DefaultVisualManager to select which node to render.

damienmarchal committed:
Merge remote-tracking branch 'upstream/master' into pr-add-link-to-defaultvisualmanagerloop

guparan commented on SofaKernel/modules/SofaSimulationCore/src/sofa/simulation/DefaultVisualManagerLoop.h:
```diff
@@ -87,9 +87,10 @@ class SOFA_SIMULATION_CORE_API DefaultVisualManagerLoop : public sofa::core::vis
         return obj;
     }
 
+    SingleLink< DefaultVisualManagerLoop, simulation::Node, BaseLink::FLAG_STOREPATH> l_node;
```
Please add a Doxygen comment `//< blablabla`

guparan commented on SofaKernel/modules/SofaSimulationCore/src/sofa/simulation/DefaultVisualManagerLoop.cpp:
```diff
@@ -40,9 +40,9 @@ int DefaultVisualManagerLoopClass = core::RegisterObject("The simplest Visual Lo
 DefaultVisualManagerLoop::DefaultVisualManagerLoop(simulation::Node* _gnode)
     : Inherit()
-    , gRoot(_gnode)
+    , l_node(initLink("targetNode","Link to the topology relevant for this object"))
```
Link to the topology?

damienmarchal committed:
[SofaKernel] Clean DefaultVisualManagerLoop,

forward declarations of the link,
namespaces,
pragma once

guparan commented:
Didn't see the Windows compilation issue yesterday. Can you look at it @damienmarchal ?

damienmarchal commented:
It is an include problem that need to be fixed. I will look at it.

damienmarchal committed:
[SofaSimulationCore] I love fixing VS specific issues.

damienmarchal commented:
[ci-build]

damienmarchal committed:
[all] Rename the forwarding helper in sofa::simulation to use the "official way" done in sofa::core

To reduce the #include drama it was decided to use forward declaratin as much
as possible combined with explicit up-casting and down-casting.

There was such mecanism in simulation/fwd.h.
When we decided to use the same mecanism in core/fwd.h some unified convention where selected.

This PR updates simulation/fwd.h so to match the selected convention and fix the rest of the code.

damienmarchal commented:
[ci-build][with-scene-tests][with-regression-tests]

damienmarchal committed:
Merge remote-tracking branch 'upstream/master' into pr-add-link-to-defaultvisualmanagerloop
